subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c

```diff
@@ -1110,376 +644,465 @@ static void fota_evt_handler(struct azure_fota_event *fota_evt)
+			LOG_ERR("Failed to get twin patch topic, az error: %d", err);
+			return -EFAULT;
 		}
+		break;
+	}
+	case AZURE_IOT_HUB_TOPIC_TWIN_REQUEST: {
+		char request_id[20];
+		char *request_id_ptr = msg->request_id.ptr;
+		size_t request_id_len = msg->request_id.size;
+		az_span request_id_span;
+
+		if (msg->request_id.size == 0) {
+			err = request_id_create_and_get(request_id, sizeof(request_id));
+			if (err) {
+				LOG_ERR("Failed to create request ID");
+				return -ENOMEM;
+			}
 
-		if ((fds[0].revents & POLLHUP) == POLLHUP) {
-			LOG_DBG("Socket error: POLLHUP");
-			LOG_DBG("Connection was unexpectedly closed");
-			break;
+			request_id_ptr = request_id;
+			request_id_len = strlen(request_id);
 		}
 
-		if ((fds[0].revents & POLLERR) == POLLERR) {
-			LOG_DBG("Socket error: POLLERR");
-			LOG_DBG("Connection was unexpectedly closed");
-			break;
+		request_id_span = az_span_create(request_id_ptr, request_id_len);
+
+		err = az_iot_hub_client_twin_document_get_publish_topic(
+			&iot_hub_client,
+			request_id_span,
+			topic,
+			sizeof(topic),
+			&topic_len);
+		if (az_result_failed(err)) {
+			LOG_ERR("Failed to create get twin topic, az error: %d", err);
+			return -EFAULT;
 		}
+
+		break;
+	}
+	default:
+		LOG_ERR("No topic specified");
+		return -ENOMSG;
 	}
 
+	param.message.topic.topic.size = topic_len;
+	param.message.topic.topic.utf8 = topic;
 
-	/* Upon a socket error, disconnect the client and notify the
-	 * application. If the client has already been disconnected this has
-	 * occurred via an MQTT DISCONNECT event and the application has
-	 * already been notified.
-	 */
-	if (!connection_state_verify(STATE_INIT)) {
-		struct azure_iot_hub_evt evt = {
-			.type = AZURE_IOT_HUB_EVT_DISCONNECTED
-		};
+	return mqtt_helper_publish(&param);
+}
 
-		/* We need to call disconnect here to ensure that the MQTT socket is closed. */
-		azure_iot_hub_disconnect();
-		azure_iot_hub_notify_event(&evt);
+int azure_iot_hub_disconnect(void)
+{
+	int err;
+
+	if (!iot_hub_state_verify(STATE_CONNECTED)) {
+		LOG_WRN("Azure IoT Hub is not connected");
+		return -ENOTCONN;
+	}
+
+	iot_hub_state_set(STATE_DISCONNECTING);
+
+	err = mqtt_helper_disconnect();
+	if (err) {
+		LOG_ERR("Failed to gracefully disconnect from MQTT broker, error: %d", err);
+		iot_hub_state_set(STATE_DISCONNECTED);
+		return -ENXIO;
 	}
 
-	/* Always revert to the initialization state if the socket has been
-	 * closed.
+	/* The MQTT library only propagates the MQTT_DISCONNECT event
+	 * if the call to mqtt_disconnect() is successful. In that case the
+	 * setting of STATE_DISCONNECTED is carried out in the mqtt_evt_handler.
 	 */
-	connection_state_set(STATE_INIT);
-	goto start;
+
+	return 0;
 }
 
-K_THREAD_DEFINE(azure_connection_poll_thread, CONFIG_AZURE_IOT_HUB_STACK_SIZE,
-		azure_iot_hub_run, NULL, NULL, NULL,
-		K_LOWEST_APPLICATION_THREAD_PRIO, 0, 0);
+int azure_iot_hub_method_respond(struct azure_iot_hub_result *result)
+{
+	int err;
+	ssize_t len;
```

Comment:
Should  this be `size_t`? `az_iot_hub_client_methods_response_get_publish_topic()` accepts `size_t *` pointer.

Reply:
Changed to `size_t`